<a href="https://colab.research.google.com/github/iammariyas/Summer_Practice/blob/main/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_2_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Задание 2.5 NumPy Сравнение эффективности

Вариант 20

Написать программу, реализующую алгоритм из индивидуального задания с использованием `list`, замерить время выполнения и сравнить с готовой реализацией алгоритма из библиотеки `NumPy`.

Для более полного анализа проводить сравнения на разных размерностях входных данных. Например, для матриц можно взять три варианта реализации: $10 × 10$, $100 × 100$, $500 × 500$.

(В зависимости от задания время выполнения алгоритма на больших размерностях может быть очень большим, в этом случае размерности можно уменьшить. И наоборот: при необходимости можно увеличить размерность или повторить действия несколько раз и усреднить результаты)

20. Найти наиболее часто встречающееся значение в матрице.

Вычислим время выполнения индивидуального задания с использованием `list`:

In [ ]:
import time
from collections import Counter
import random

In [ ]:
def most_value(matrix):
  lst = []
  for row in matrix:
    lst.extend(row)

  cnt = Counter(lst)
  max_cnt = max(cnt.values())
  repeat = [key for key, value in cnt.items() if value == max_cnt]
  return repeat[0] if len(repeat) == 1 else repeat

In [ ]:
matrix = [[random.randint(1, 100) for _ in range(10)] for _ in range(10)]
%timeit most_value(matrix)

22.2 µs ± 1.37 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [ ]:
matrix = [[random.randint(1, 100) for _ in range(100)] for _ in range(100)]
%timeit most_value(matrix)

565 µs ± 51.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [ ]:
matrix = [[random.randint(1, 100) for _ in range(500)] for _ in range(500)]
%timeit most_value(matrix)

20.7 ms ± 5.98 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


Теперь узнаем время выполнения индвидуального задания с использованием готовой реализации алгоритма из библиотеки `NumPy`

In [ ]:
import numpy as np

In [ ]:
def np_most_value(matrix):
  matrix_rvl = matrix.ravel()
  el, cnt = np.unique(matrix_rvl, return_counts=True)
  max_cnt = np.max(cnt)
  repeat = el[cnt == max_cnt]
  return repeat[0] if len(repeat) == 1 else repeat

In [ ]:
matrix = np.random.randint(1, 100, (10, 10))
%timeit np_most_value(matrix)

44.5 µs ± 10.6 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [ ]:
matrix = np.random.randint(1, 100, (100, 100))
%timeit np_most_value(matrix)

161 µs ± 51.9 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [ ]:
matrix = np.random.randint(1, 100, (500, 500))
%timeit np_most_value(matrix)

2.85 ms ± 926 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Теперь проанализируем результаты и сравним время выполнения функций `most_value` и `np_most_value`:

Рассмотрим результаты, когда имеется матрица $10 × 10$:

Функция `np_most_value` $44.5 µs µs ÷ 22.2 µs ≈ 2 $ раза медленнее выполняется. Это обусловлено тем, что `NumPy`на маленьких данных подготовительные действия библиотеки (проверка типа данных, выделить память под массивы, запускает оптимизированные функции) занимают больше времени, чем работа алгоритма. Игра не стоит свеч, поэтому на маленьких данных лучше использовать методы из встроенных библиотек.

Рассмотрим результаты, когда имеется матрица $100 × 100$:

`NumPy` в $565 µs ÷ 161 µs ≈ 3.5$ раз быстрее стандартных модулей Python. Мы начинаем видеть преимущества использования библиотеки NumPy.

Рассмотрим результаты, когда имеется матрица $500 × 500$:

NumPy выполняется в $20.7 ms÷2.85 ms≈ 7.3$ раз быстрее.